kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/records/RecordStream.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption.records;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+
+import org.apache.kafka.common.record.MemoryRecords;
+import org.apache.kafka.common.utils.ByteBufferOutputStream;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * <p>An API for manipulating batches of {@link org.apache.kafka.common.record.Record}s.
+ * This is loosely inspired by {@code java.util.stream.Stream},
+ * but more sympathetic to Kafka's {@link org.apache.kafka.common.record.RecordBatch}
+ * and {@link org.apache.kafka.common.record.Record} APIs.</p>
+ *
+ * <p>Conceptually a {@code RecordStream} is an ordered sequence of {@link org.apache.kafka.common.record.RecordBatch}es
+ * with the ability to associate some state with records.</p>
+ *
+ * <h2>Map-like operations</h2>
+ * <p>A stream created from {@link #ofRecords(MemoryRecords)} has no state associated with the records.
+ * Methods are provided to map the state of streams</p>
+ * <dl>
+ * <dt>{@link #mapConstant(Object)}</dt>
+ *   <dd>associates the same state with each record in the stream.</dd>
+ * <dt>{@link #mapPerRecord(RecordMapper)}</dt>
+ *   <dd>associates per-record state with each record in the stream.</dd>
+ * </dl>
+ *
+ * <h2>Collect-like operations</h2>
+ * <p>It is also possible to transform a stream to some non-stream representation:</p>
+ * <dl>
+ * <dt>{@link #toSet(RecordMapper)}</dt>
+ *   <dd>to simultaneously map and convert a stream to a set.
+ *       This is like {@link java.util.stream.Collectors#toMap(Function, Function)}.</dd>
+ * <dt>{@link #toMemoryRecords(ByteBufferOutputStream, RecordTransform)}</dt>
+ *   <dd>Simultaneously map and convert a stream to a {@link MemoryRecords}</dd>
+ * </dl>
+ * @param <T> The type of associated state.
+ */
+public class RecordStream<T> {
```

Comment:
Not entirely sure what I feel about exposing a T. 

It sort of feels to me that RecordStream is really an interface and each PairType is a different implementation, but I doubt I've absorbed enough detail about it yet to see how realistic that is. 

Currently we have 
| Method  | return type | proposal |
| ------------- | ------------- | ------------- |
| `RecordStream.ofRecordsWithIndex`  | `RecordStream<Integer>` | `IndexedRecordStream` | 
| `RecordStream.ofRecords`   | `RecordStream<Void> `  | `RawRecordStream`* |
| `RecordStream.mapPerRecord`   | `RecordStream<?> `  | `??RecordStream`** |


Admittedly the final row is an instance method not a factory...  
(*) Naming needs love
(**) This is what makes me think I'm missing something.

Reply:
I think https://github.com/tombentley/kproxy/pull/14 covers this.

Reply:
I merged tombentley#14, so resolving this one.